### PR TITLE
Implement Dump method on part of DynamicTypeHandler hierarchy

### PR DESCRIPTION
**Pull request by ricobbe**
As part of my current investigation, I need to be able to dump a human-readable represntation of DictionaryTypeHandlerBase objects.  Add a virtual Dump method to DynamicTypeHandle to allow this, and provide stub implementations for other subtypes of DynamicTypeHandler.  (As the need arises, I may provide full implementations for these stubs; others are welcome to do so if they get to it before I do.)

### CODE REVIEW

File: lib/Runtime/Base/PropertyRecord.cpp [hunk 1]
```diff
@@ -134,6 +134,22 @@ namespace Js
         }
     }
 
+#if DBG_DUMP
+	void PropertyRecord::Dump(unsigned indent) const
```
Comment (obastemur) [inline lib/Runtime/Base/PropertyRecord.cpp]: nit: space

File: lib/Runtime/Types/DictionaryTypeHandler.cpp [hunk 2]
```diff
@@ -2909,6 +2909,41 @@ namespace Js
     }
 #endif
 
+#if DBG_DUMP
+	template<typename T> void DictionaryTypeHandlerBase<T>::Dump(unsigned indent) const {
```
Comment (obastemur) [inline lib/Runtime/Types/DictionaryTypeHandler.cpp]: nit: padding

File: lib/Runtime/Types/DictionaryTypeHandler.cpp [hunk 3]
```diff
@@ -2909,6 +2909,41 @@ namespace Js
...
+        const auto padding(_u(""));
+        const unsigned fieldIndent(indent + 2);
+        const unsigned mapLabelIndent(indent + 4);
+        const unsigned mapValueIndent(indent + 6);
+
+        Output::Print(_u("%*sDictionaryTypeHandlerBase (0x%p):\n"), indent, padding, this);
+        if (this->propertyMap == nullptr)
+        {
+            Output::Print(_u("%*spropertyMap: null\n"), fieldIndent, padding);
```
Comment (obastemur) [inline lib/Runtime/Types/DictionaryTypeHandler.cpp]: nit: `<null>` instead? to be similar with others?

**Comment by obastemur** [comment]
I've been looking into Type system a bit recently. I'm glad to see this is coming.

couple of nits.

LGTM

**Comment by obastemur** [comment]
BTW; about the Clang CI errors. Try casting to `(void*)` i.e. `(void*)this->propertyMap`

File: lib/Runtime/Base/PropertyRecord.cpp (hunk 1 above)
Comment (ricobbe) [reply lib/Runtime/Base/PropertyRecord.cpp]: Fixed.  (My emacs config was incorrectly not indenting definitions inside a namespace -- holdover from previous job -- so I was doing it by hand via C-q C-i, also a holdover from the previous job.  Thx for the comment; this motivated me to fix the emacs config to indent within a namespace!)

File: lib/Runtime/Types/DictionaryTypeHandler.cpp (hunk 2 above)
Comment (ricobbe) [reply lib/Runtime/Types/DictionaryTypeHandler.cpp]: Fixed

File: lib/Runtime/Types/DictionaryTypeHandler.cpp (hunk 3 above)
Comment (ricobbe) [reply lib/Runtime/Types/DictionaryTypeHandler.cpp]: Fixed

File: lib/Runtime/Types/DictionaryTypeHandler.cpp [hunk 4]
```diff
@@ -2909,6 +2909,41 @@ namespace Js
...
+
+        Output::Print(_u("%*sDictionaryTypeHandlerBase (0x%p):\n"), indent, padding, this);
+        if (this->propertyMap == nullptr)
+        {
+            Output::Print(_u("%*spropertyMap: <null>\n"), fieldIndent, padding);
+        }
+        else
+        {
+            Output::Print(_u("%*spropertyMap: 0x%p\n"), fieldIndent, padding, static_cast<void*>(this->propertyMap));
+            for (auto iter = this->propertyMap->GetIterator(); iter.IsValid(); iter.MoveNext())
```
Comment (digitalinfinity) [inline lib/Runtime/Types/DictionaryTypeHandler.cpp]: Can you use Map here instead of using an iterator?

Comment (ricobbe) [reply lib/Runtime/Types/DictionaryTypeHandler.cpp]: Fixed.  I'd overlooked Map when reading through BaseDictionary; thanks!

**Pull request merged**